lib/pos_mempool.go

```diff
@@ -0,0 +1,417 @@
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, dmp.latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check transaction signature
+	if _, err := dmp.latestBlockView.VerifySignature(txn, uint32(dmp.latestBlockHeight)); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Signature validation failed")
+	}
+
+	// If we get this far, it means that the transaction is valid. We can now add it to the mempool.
+	// We lock the mempool to ensure that no other thread is modifying it while we add the transaction.
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	// Add the transaction to the mempool and then prune if needed.
+	if err := dmp.addTransactionNoLock(mempoolTx, true); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	if err := dmp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) addTransactionNoLock(txn *MempoolTx, persistToDb bool) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+	txnFee := txn.Tx.TxnFeeNanos
+
+	// Validate that the user has enough balance to cover the transaction fees.
+	spendableBalanceNanos, err := dmp.latestBlockView.GetSpendableDeSoBalanceNanosForPublicKey(userPk.ToBytes(),
+		uint32(dmp.latestBlockHeight))
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem getting spendable balance")
+	}
+	if err := dmp.ledger.CanIncreaseBalance(*userPk, txnFee, spendableBalanceNanos); err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem checking balance increase for transaction with"+
+			"hash %v, fee %v", txn.Tx.Hash(), txnFee)
+	}
+
+	// If we get here, it means that the transaction's sender has enough balance to cover transaction fees. We can now
+	// add the transaction to mempool.
+	if err := dmp.txnRegister.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem adding txn to register")
+	}
+
+	// We update the reserved balance to include the newly added transaction's fee.
+	dmp.ledger.IncreaseBalance(*userPk, txnFee)
+
+	// Emit an event for the newly added transaction.
+	if persistToDb {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventAdd,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// RemoveTransaction is the main function for removing a transaction from the mempool.
+func (dmp *PosMempool) RemoveTransaction(txnHash *BlockHash) error {
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.RemoveTransaction: ")
+	}
+
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	// Get the transaction from the register.
+	txn := dmp.txnRegister.GetTransaction(txnHash)
+	if txn == nil {
+		return nil
```

Comment:
Should this be an error if we try to remove a transaction that isn't found in the register? jw

Reply:
Returning nil here is the desired behavior. When successfully processing a block, the server will try to naively remove the block's txns from the mempool to ensure that a connected txn is no longer in the mempool. So as long as a txn isn't in the mempool at the end of this operation, we consider the remove a success.